### PR TITLE
Remove dependecy toward libsndfile on all OS

### PR DESCRIPTION
**Pull request by mantognini**
This implements #604. It's a rebase of `feature/no_libsndfile_osx` onto master (which includes `feature/no_libsndfile`) so be extra careful when proofreading it. Hopefully I haven't made any mistake but there were a lot of conflicts so... 

What's done & what's to be done:

 * [x] Android (all libs included)
 * [x] Android – make loading sound compatible with APK
 * [x] iOS (all libs included)
 * [x] Linux (no binary to be provided)
 * [x] OS X (all libs included)
 * [x] Window (all libs included)

It also includes the base step toward #620.

### CODE REVIEW

**Comment by mantognini** [comment]
@MarioLiebisch Just to be sure, there's absolutely no difference in `config_types.h` across all different supported archs (mips, armeabi, armeabi-v7a, x86)?

Assuming it is the case, then:

@LaurentGomila (and others): until now we had `extlibs/headers/libsndfile/{android,osx,windows}`. Should we also have `extlibs/headers/ogg/{android,ios,osx,windows}` to follow the same spirit or do we just don't care and add `config_types.h` directly in `extlibs/headers/ogg`? Alternatively we could edit `os_types.h` to contain android default value (with the proper `#elif` statement). What is your opinion on the subject?

**Comment by LaurentGomila** [comment]
If it's the same then there's no reason to duplicate it per platform. There's no "spirit", let's just be pragmatic ;)

**Comment by MarioLiebisch** [comment]
From what I can tell, the header file is exactly the same for all Android architectures (compared armeabi, x86, and x86_64) as long as we're forcing GCC to be used.

Not sure regarding other platforms and/or compilers though.

**Comment by mantognini** [comment]
> If it's the same then there's no reason to duplicate it per platform.

Alright. So I've checked and got the same file on OS X than @MarioLiebisch on Android. I don't know about iOS, Linux or Windows (@eXpl0it3r, since you've build them, do you know?).

For reference, `config_types.h` for Android is [here](https://github.com/MarioLiebisch/SFML/commit/3b05ee110c1f029fe26b92eb532380cba1c639d8#diff-0).

**Comment by eXpl0it3r** [comment]
I've only built the FLAC libraries which didn't require any `config_types.h`. The OGG libs etc. were built by @LaurentGomila.

**Comment by mantognini** [comment]
Right, sorry about that. So, Laurent? ;)

**Comment by mantognini** [comment]
I just checked on Linux and the file is different. Don't know about Windows.

Even though we don't provide Linux binaries, it's not that hard to make a special rules for Android in `src/SFML/Audio/CMakeLists.txt`.

**Comment by mantognini** [comment]
@MarioLiebisch I've updated (rebased + force pushed) the branch, can you check that it now build fine on Android? Thanks.

**Comment by MarioLiebisch** [comment]
Okay, this just needs a minor fix for the paths (the way you did it won't work since the relative path (`ogg/config_types.h`) won't work. I'm fixing this with the next commit I push right after this comment.

This leaves just the minor issue of the Android version trying to read the audio files from the file system rather than the assets. I've modified `FileInputStream` for testing and it works, but I think we should make `ResourceStream` a public `AssetInputStream` or `RessourceInputStream` to allow users to pick either one. SFML funtions taking paths should use that one by default I'd say.

What do you think?

**Comment by mantognini** [comment]
> Okay, this just needs a minor fix for the paths (the way you did it won't work since the relative path (`ogg/config_types.h`) won't work. I'm fixing this with the next commit I push right after this comment.

Right!

> I've modified FileInputStream for testing and it works, but I think we should make `ResourceStream` a public `AssetInputStream` or `RessourceInputStream` to allow users to pick either one. SFML funtions taking paths should use that one by default.

I'm not sure to fully grok how things work on Android -- I personally trust your judgment. ;-)

**Comment by MarioLiebisch** [comment]
I'm not sure to fully grok how things work on Android -- I personally trust your judgment. ;-)

Basically you've got two file system. One inside your APK's assets directory (whatever is in the zip archive; no special permissions required) and your typical file system root (requires permission).

**Comment by mantognini** [comment]
Some kind of sandboxing if I understand correctly?

Do you think it deserve a dedicated PR? If yes, this one should be ready for review & merge.

**Comment by MarioLiebisch** [comment]
Yes, consider it a readonly sandboxed/bundled file system (pretty much like Java JARs).

The current version in this branch changes SFML's existing behavior (2.2):

On Android all SFML classes will load from assets only, except sound, which would load from the native filesystem (until fixed/expanded).

**Comment by mantognini** [comment]
> The current version in this branch changes SFML's existing behavior (2.2)

Ho, okay. I didn't get that. Then it makes sense to fix it in there too.

**Comment by MarioLiebisch** [comment]
I have a modified *FileInputStream.cpp*, which essentially restores the old behavior. We could use that if you want to get this PR done, but we'd have to revert those changes later on again (when moving the asset reading to its own class).

**Comment by LaurentGomila** [comment]
Managing resource files on Android is an unrelated task, please don't do it in this PR, it is already big enough :grin:

**Comment by MarioLiebisch** [comment]
@LaurentGomila So let's go with the fix (have it ready, just have to commit/push it), then clean that up later on? Or leave it broken intentionally just for this PR/merge?

**Comment by LaurentGomila** [comment]
What exactly is broken? As far as I know, file loading is working fine on Android. If you want to change some behaviour then it should definitely not be done here.

**Comment by MarioLiebisch** [comment]
Right now, this PR uses standard streams for  file loading only, which won't access an APK's asset files on Android (therefore breaking existing code).

**Comment by LaurentGomila** [comment]
> This PR introduces a new abstraction layer InputStream, to read files.

I had totally forgotten about that part :smile:

> modify FileInputStream to provide the old behavior.

Ok, let's go for this, and don't hesitate to open a new task or submit a PR if you feel like things could be improved for Android.

**Comment by mantognini** [comment]
Alright. Let us know when you have something ready. BTW, I've added this sub-task to the description above.

**Comment by MarioLiebisch** [comment]
Android should work fine now (example app runs fine for me).

However, the precompiled libraries for MinGW are obviously missing *vorbisenc*. I removed the dependency to try building and got the following error using x86 GCC 4.9.2:

    Linking CXX shared library ..\..\..\lib\sfml-audio-2.dll
    ../../../../../../extlibs/libs-mingw/x86/libFLAC.a: error adding symbols: Archive has no index; run ranlib to add one
    collect2.exe: error: ld returned 1 exit status

There seems to be something amiss here?

**Comment by mantognini** [comment]
I'll need to rebase this one to match latest master history. In the meantime, can someone build vorbisenc for windows? ( @LaurentGomila maybe? )

**Comment by eXpl0it3r** [comment]
I can give vorbisenc a shot this evening.

**Comment by mantognini** [comment]
Would be great! :-)

**Comment by phoenixx2** [comment]
I'm fairly new to Git, so I'm not entirely sure if this has been fixed, but In MemoryInputStream.cpp:

Int64 count = endPosition <= m_size ? size : size - m_offset;

should be:

Int64 count = endPosition <= m_size ? size : m_size - m_offset;

Otherwise end of files stored in memory are not properly detected.

**Comment by mantognini** [comment]
I guess you're right but could someone else confirm this?

**Comment by LaurentGomila** [comment]
Yes, seems like you're right :+1:

**Comment by mantognini** [comment]
There, it's fixed. When I've got the lib for Windows I'll squash the last commit where appropriate and update the branch.

**Comment by eXpl0it3r** [comment]
I hope this [http://my-gate.net/pub/libvorbisenc.a](libvorbisenc.a) works..

**Comment by eXpl0it3r** [comment]
@mantognini :neckbeard:

**Comment by mantognini** [comment]
Yes, as soon as I've updated the branch I let you know so you can test. ;)

**Comment by eXpl0it3r** [comment]
Just tried it and noticed that it doesn't work. Besides I totally forgot to build an x64 version and one for VS. :wink:

**Comment by phoenixx2** [comment]
Has anyone been able to compile things in a Release build on Windows?  I'm getting the following error in VS 2010, even after a clean re-build:
error C1047: The object or library file '.\SFML\extlibs\libs-msvc\x86\ogg.lib' was created with an older compiler than other objects; rebuild old objects and libraries

No such error exists in Debug build.  I also noticed that /LTG is required to be specified once linking to these pre-built libraries.  I believe removing /LTG from the libraries being pre-built will solve this issue based on my understanding of this article:
https://msdn.microsoft.com/en-us/library/ms173554%28v=vs.90%29.aspx

As a side note, I wasn't able to link against everything for FLAC support either, getting this error (I just hard-coded removed FLAC on my end) :
error LNK2001: unresolved external symbol __imp__FLAC__stream_encoder_init_file
(8 unresolved externals in total, all FLAC)

I am linking to the following libs:
openal32.lib
ogg.lib
vorbis.lib
vorbisfile.lib
flac.lib

**Comment by eXpl0it3r** [comment]
So I've rebuild all the new libraries for both x86 and x64 with MinGW and it seems to work fine. However VS still seems to be tricky, mostly because there's no `libvorbisenc` to be built and I get again the previously mentioned `flac.lib(stream_decoder.obj) : error LNK2001: Unresolved external symbol"_fopen_utf8".` and I don't really know how to resolve it or would that now need `libvorbisenc`?

**Comment by MarioLiebisch** [comment]
Current HEAD builds fine for me in MSVC 2013 as release, ignoring some warnings:

    SFML/System/FileInputStream.hpp(54): warning C4275: non dll-interface class 'sf::InputStream' used as base for dll-interface class 'sf::FileInputStream'
        System/InputStream.hpp(41) : see declaration of 'sf::InputStream'
        include\SFML/System/FileInputStream.hpp(53) : see declaration of 'sf::FileInputStream'

    SFML\Audio\SoundFileReaderFlac.cpp(281): warning C4244: 'argument' : conversion from 'sf::Uint64' to '__w64 int', possible loss of data

    vorbis.lib(info.obj) : MSIL .netmodule or module compiled with /GL found; restarting link with /LTCG; add /LTCG to the link command line to improve linker performance

Trying MinGW next.

**Comment by eXpl0it3r** [comment]
It shouldn't if you do a full rebuild with CMake, because it will fail in CMake due to vorbisenc missing.

**Comment by LaurentGomila** [comment]
My conclusion regarding ```_fopen_utf8``` is that it is a mistake on Windows, so to compile the library I had to change something directly in a header (comment or rename something, so that the macros end up calling the regular fopen).

**Comment by eXpl0it3r** [comment]
> My conclusion regarding _fopen_utf8 is that it is a mistake on Windows, so to compile the library I had to change something directly in a header (comment or rename something, so that the macros end up calling the regular fopen).

Na, it's all there, one just would need to link against `win_utf8_io`, since that's just one header file and source file, I now included them directly into libflac and it now works. Not sure if we really want to be that hacky, but the other option is to ship and link yet another library. :wink:

**Comment by MarioLiebisch** [comment]
So it would compile with the bundled precompiled library but not with another copy compiled from source? Doesn't sound very reliable and would avoid it.

**Comment by eXpl0it3r** [comment]
Well we can also add it, but it's only something that needs to be linked on Windows and it's essentially an utility library.

I've no idea how to properly adjust CMake...

**Comment by LaurentGomila** [comment]
Let's keep it simple. And on Windows, nobody will recompile SFML with another copy than the one we provide.

**Comment by eXpl0it3r** [comment]
Okay, I've [rebuilt everything](http://my-gate.net/pub/extlibs.7z) and it "works".

MinGW builts perfectly fine, no issues as far as I can tell.
Visual Studio links against a certain version of the runtime libs and as such you get conflicts when using different settings (i.e. debug vs release). I've specified `/TC` for all the libs, but it didn't seem to have helped. I'm not even sure which library creates the conflict.

Besides the warning, it still seems to run fine.

**Comment by mantognini** [comment]
Good! I'll update the branch later today.

**Comment by LaurentGomila** [comment]
For Visual C++ you have to set a special option, but I don't remember which one :smile:

**Comment by MarioLiebisch** [comment]
`/NOSTDLIBS` or whatever it's called?

**Comment by LaurentGomila** [comment]
Yes, most likely :)

**Comment by eXpl0it3r** [comment]
You probably meant `/NODEFAULTLIB` and nope that option didn't remove the conflict.

**Comment by LaurentGomila** [comment]
I think it's /Zl (Omit Default Library Name).

**Comment by mantognini** [comment]
I'm not familiar with those so should I change something in the cmake script? If yes, where and how?

**Comment by eXpl0it3r** [comment]
@LaurentGomila That seems to have been it!
@mantognini No that's just for building the dependencies.

**Comment by mantognini** [comment]
@eXpl0it3r so can I use your extlibs.7z as is or should I wait for a new version?

**Comment by eXpl0it3r** [comment]
I've replaced the other file with the new version](http://my-gate.net/pub/extlibs.7z). This should now not produce any kind of linker error.

Additionally I've noticed some warnings in the new classes that I'll investigate a bit later.

Also in the Android extlibs directory there's `libFLAC++.a` but we don't use `libFLAC++` as such I think they can get delete.

**Comment by eXpl0it3r** [comment]
I'm getting these warnings:

> SoundFileFactory.cpp
SFML\include\SFML/System/FileInputStream.hpp(54) : warning C4275: non – DLL-interface 'sf::InputStream' used as base for DLL-interface 'sf::InputStream'
f::FileInputStream' used
    SFML\include\SFML/System/InputStream.hpp(41): See declaration of 'sf::InputStream'
    SFML\include\SFML/System/FileInputStream.hpp(53): See declaration of 'sf::FileInputStream'

> SoundFileReaderOgg.cpp
SFML\include\SFML/System/MemoryInputStream.hpp(44) : warning C4275: non – DLL-interface 'sf::InputStream' used as base for DLL-interface 'sf::InputStream'
'sf::MemoryInputStream' used
    SFML\include\SFML/System/InputStream.hpp(41): See declaration of 'sf::InputStream'
    SFML\include\SFML/System/MemoryInputStream.hpp(43): See declaration of 'sf::MemoryInputStream'
        
> SoundFileReaderFlac.cpp
SFML\src\SFML\Audio\SoundFileReaderFlac.cpp(281) : warning C4244: 'Argument': Converting from 'sf::Uint64' to '__w64 int', possible data lose
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xutility(2132) : warning C4996: 'std::_Copy_impl': Function call with parameters that may be unsafe - this call relies on the caller to check
that the passed values are correct. To disable this warning, use -D_SCL_SECURE_NO_WARNINGS. See documentation on how to use Visual C++ 'Checked Iterators'
    C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xutility(2113): See declaration of 'std::_Copy_impl'
    SFML\src\SFML\Audio\SoundFileReaderFlac.cpp(280): See reference of the instancing of the compiled function-template "_OutIt std::copy<std::_Vector_iterator<
    
>    std::_Vector_val<std::_Simple_types<short>>>,sf::Int16*>(_InIt,_InIt,_OutIt)".
    with
    [
        _OutIt=sf::Int16 *,
        _InIt=std::_Vector_iterator<std::_Vector_val<std::_Simple_types<short>>>
    ]

The first error must have something to do with `dllexport()` not sure what exactly. And the second seems to be a simple casting issue.

**Comment by LaurentGomila** [comment]
For the first one, add ```SFML_SYSTEM_API``` there https://github.com/LaurentGomila/SFML/blob/feature/no_libsndfile_all_os/include/SFML/System/InputStream.hpp#L40

**Comment by MarioLiebisch** [comment]
> The first error must have something to do with dllexport() not sure what exactly. And the second seems to be a simple casting issue.

As far as I know it's basically a warning that you're exporting a derived class, but not it's base class.

**Comment by mantognini** [comment]
The branch has been updated. Could you give it a go on Linux, Windows, iOS and Android?

**Comment by LaurentGomila** [comment]
By the way, could someone remove all the old branches (I have no idea how to remove a remote branch)? It's becoming a mess.

**Comment by MarioLiebisch** [comment]
Android is building and running fine for me.

@LaurentGomila https://github.com/LaurentGomila/SFML/branches There are the buttons you're looking for. :)

**Comment by LaurentGomila** [comment]
This is the first time I ever see this page, thanks :open_mouth:

> Android is building and running fine for me.

Have you tested all the supported audio formats?

**Comment by mantognini** [comment]
> By the way, could someone remove all the old branches 

no_libsndfile and no_libsndfile_osx were removed a few days ago. The remaining branches are still unmerged I think.

**Comment by MarioLiebisch** [comment]
> Have you tested all the supported audio formats?

Meh, isn't it that easy? :) Have to try again tomorrow then.

**Comment by LaurentGomila** [comment]
Configures, builds and runs fine on Windows / Visual Studio 2013 :+1:

> Meh, isn't it that easy? :) Have to try again tomorrow then.

Then when you say "running fine", running what? :stuck_out_tongue:

**Comment by eXpl0it3r** [comment]
I don't see the change [Laurent stated](https://github.com/LaurentGomila/SFML/pull/757#issuecomment-75547446) for fixing the warnings I posted.

**Comment by MarioLiebisch** [comment]
> Then when you say "running fine", running what? :P

The example compiled and installed with its bundled bird sounds.

**Comment by mantognini** [comment]
> I don't see the change Laurent stated for fixing the warnings I posted.

Fixed

**Comment by eXpl0it3r** [comment]
Not really that important, but I wonder, can we mark the `_INCLUDE_DIR` and `_LIBRARY` entries as advanced somehow?
Because with a few more dependencies CMake now feels quite cluttered.

CMake

**Comment by LaurentGomila** [comment]
http://www.cmake.org/cmake/help/v3.0/command/mark_as_advanced.html

**Comment by mantognini** [comment]
@eXpl0it3r if you wish to change something, either send me a patch or push to this branch -- I won't be able to work on it before next week and I don't want to stop the progress of this feature now. ;-)

**Comment by MarioLiebisch** [comment]
Already adjusted the files. Only visible non-advanced library is GLEW now. Could update that as well, but that's kinda off-topic here? Opinions?

**Comment by eXpl0it3r** [comment]
Since the new libraries get introduced with this PR it also makes sense to add it to this branch and thus it's not really off-topic.
Anyways, I'll have to update the VS binaries since VS 2010 doesn't support SSE2 and thus I need to explicitly disable it when building with VS 2013...

**Comment by MarioLiebisch** [comment]
I'm talking about *FindGLEW*, not Vorbis/Ogg/FLAC.

**Comment by eXpl0it3r** [comment]
You can ignore GLEW, since it will be removed later on anyways.

**Comment by LaurentGomila** [comment]
While you're in the CMake modules, FindSndfile.cmake can be removed here :)

**Comment by eXpl0it3r** [comment]
Well no matter what I do, it doesn't want to link with VS 2010 x64. I think it's more of a toolchain issue than something with the libraries though...
I get undefined references to `__vdecl_cos2` which Google doesn't even have an entry for...

**Comment by MarioLiebisch** [comment]
Sounds like some implementation part of the math library. Are you sure there's no library conflict? Does it happen with other versions of MSVC or can't you test those?

**Comment by eXpl0it3r** [comment]
As far as I could test it builds fine with VS 2012 and 2013 as well as VS 2010 x86...
The x64 has always been more of a hack...

**Comment by eXpl0it3r** [comment]
As far as I could test it builds fine with VS 2012 and 2013 as well as VS 2010 x86...
The x64 has always been more of a hack...

**Comment by MarioLiebisch** [comment]
Might indeed be some toolchain bug then. Odd that it doesn't happen for x86.

**Comment by eXpl0it3r** [comment]
As far as I could test it builds fine with VS 2012 and 2013 as well as VS 2010 x86...
The x64 has always been more of a hack...

**Comment by MarioLiebisch** [comment]
Would leave it as a toolchain bug for now then.

**Comment by LaurentGomila** [comment]
One more thing before validating this branch: someone should check on iOS. I probably won't be able to do it :(

**Comment by phoenixx2** [comment]
VS 2010 here, updated libs, but in addition to FLAC not working, OGG is broken now too:

error LNK2001: unresolved external symbols:
__ltod3
__libm_sse2_log_precise
__libm_sse2_pow_precise
__libm_sse2_atan_precise
__libm_sse2_exp_precise
__libm_sse2_sqrt_precise
__libm_sse2_cos_precise
__libm_sse2_sin_precise
___vdecl_cos2
__libm_sse2_acos_precise
__dtol3

@eXpl0it3r /arch:SEE and /arch:SSE2 are both available on VS 2010, so I'm not sure what you're referring to.

**Comment by eXpl0it3r** [comment]
@phoenixx2 Yes, I've fixed these issues already. Do you have by any chance the VS 2010 x64 compiler installed?

**Comment by phoenixx2** [comment]
@eXpl0it3r As far as I know, I'm using 32bit pretty much everything. If there's a way that I'm making a 32bit app using the referenced 64bit compiler, I wouldn't really know how to check for that?

**Comment by eXpl0it3r** [comment]
You'd know if you had installed VS 2010 x64. I'll try to fix this someone otherwise I'll just provide the updated dependencies.

**Comment by eXpl0it3r** [comment]
Rebuilt everything with VS 2010 directly and it seems to work fine for VS 2010, 2012 and 2013 (x86 & x64): [libs-msvc.7z](http://my-gate.net/pub/libs-msvc.7z)

**Comment by LaurentGomila** [comment]
> Also not sure how easily you can change that

A cast to sf::Int64?

**Comment by phoenixx2** [comment]
@LaurentGomila I have a few changes on my end for this branch, that I will port over to the main branch, but I'm just wondering if it's worth putting them up now:

-WAV reading overrun bug:
last few lines should be the following, as subChunkSize can be higher than 1024 (happened to me):
// unknown chunk, skip it
Int64 offset = m_stream->tell() + subChunkSize;
if (m_stream->seek(offset) != offset)
    return false;

-wavpack (wv) read/write and mp3 read (via libmpg123)

-file range specification for creating audio from files for both SFML and CSFML
allows you to create audio from a file with a specified offset and size

-retrieving generally used file extension for a music
mostly for my purposes, doubt anyone else would benefit much (for identifying wavs to be converted to wvs)

-use __stdcall in Win32 so CSFML can work as a standard Win32 API dll
had to do this for proper support in my language of choice

-warnings
a few MSVC warnings (similar to above) when showing all errors on MSVC

@eXpl0it3r Thanks for the fix, my release builds now work, and /LTG isn't required to be specified anywhere, which is nice for faster linking!

**Comment by LaurentGomila** [comment]
@phoenixx2 only bugs and warnings are relevant here (to avoid testes continuously reporting the same issues when testing the branch), further additions will be treated once it is merged.

**Comment by LaurentGomila** [comment]
```
1>vorbisenc.lib(vorbisenc.obj) : warning LNK4099: PDB 'vc120.pdb' was not found with 'vorbisenc.lib(vorbisenc.obj)' or at 'D:\Developpement\SFML\Build\SFML\SFML\examples\sound\Debug\vc120.pdb'; linking object as if no debug info
1>vorbisfile.lib(vorbisfile.obj) : warning LNK4099: PDB 'vc120.pdb' was not found with 'vorbisfile.lib(vorbisfile.obj)' or at 'D:\Developpement\SFML\Build\SFML\SFML\examples\sound\Debug\vc120.pdb'; linking object as if no debug info
1>vorbis.lib(analysis.obj) : warning LNK4099: PDB 'vc120.pdb' was not found with 'vorbis.lib(analysis.obj)' or at 'D:\Developpement\SFML\Build\SFML\SFML\examples\sound\Debug\vc120.pdb'; linking object as if no debug info
1>vorbis.lib(bitrate.obj) : warning LNK4099: PDB 'vc120.pdb' was not found with 'vorbis.lib(bitrate.obj)' or at 'D:\Developpement\SFML\Build\SFML\SFML\examples\sound\Debug\vc120.pdb'; linking object as if no debug info
1>vorbis.lib(block.obj) : warning LNK4099: PDB 'vc120.pdb' was not found with 'vorbis.lib(block.obj)' or at 'D:\Developpement\SFML\Build\SFML\SFML\examples\sound\Debug\vc120.pdb'; linking object as if no debug info
1>vorbis.lib(codebook.obj) : warning LNK4099: PDB 'vc120.pdb' was not found with 'vorbis.lib(codebook.obj)' or at 'D:\Developpement\SFML\Build\SFML\SFML\examples\sound\Debug\vc120.pdb'; linking object as if no debug info

... and a lot more with all the audio libs ...
```

This happens when compiling an executable with static SFML libraries. Not sure why it only happens with the new dependencies. Maybe you compiled them with debug info enabled (or even worse: in debug mode :p)?

**Comment by eXpl0it3r** [comment]
Arg... why can't stuff just be simple. :unamused:

**Comment by eXpl0it3r** [comment]
I've rebuilt and [reuploaded](http://my-gate.net/pub/libs-msvc.7z) them. It links fine with VS 2013 in static mode for me...

**Comment by LaurentGomila** [comment]
> You used the new 7z archive I provided, right?

No, I always test the head of the branch.

**Comment by eXpl0it3r** [comment]
Well these are again outdated, so any comment about them is useless. :wink:

**Comment by LaurentGomila** [comment]
Ok, everything builds fine with the new libs, sorry :grin:

**Comment by mantognini** [comment]
MSVC binaries updated. Could you give it a final test?

**Comment by zsbzsb** [comment]
I just compiled this branch (clean CMake cache) and I was able to run the pong example without any issues. I compiled in release with static libs on MSVC2013 and it worked perfectly.

**Comment by LaurentGomila** [comment]
I've made the modifications needed to make the audio module work on iOS. It compiles fine but I haven't been able to test it; I can work with most IDEs but XCode is really a mystery for me... It kept on throwing linker errors about architectures, whatever I did (an example: both SFML and my test project have the exact same settings, but one creates i386 libraries for the simulator and the other wants x86_64...).

**Comment by eXpl0it3r** [comment]
Maybe @BlueCobold can pick it up.

**Comment by BlueCobold** [comment]
XCode is evil about the required libs. The emulator needs i386 and x64, however, it never wants to output these in any arm-libraries. You need to manually merge these into a fat library which can handle emulator and physical devices at the same time. I can do that, no problem.

**Comment by LaurentGomila** [comment]
No, the libs have been compiled by @mantognini and should be ok. I just failed to setup a test program for the audio module.

**Comment by BlueCobold** [comment]
Ah, are you sure your settings for target architecture and allowed architecture match up? However, if the lib really contains *all* architectures, there shouldn't be a problem no matter what settings your test program uses.

**Comment by LaurentGomila** [comment]
> Ah, are you sure your settings for target architecture and allowed architecture match up?

Yes, but don't bother trying to help me, first this is not relevant in this PR, and secondly I leave this to someone with more knowledge in Xcode and iOS development :)

> However, if the lib really contains all architectures, there shouldn't be a problem no matter what settings your test program uses.

Only the external libs are universal. The produced SFML libraries are not, they are compiled for armv7 if the target is iOS device, and i386 and/or x86_64 if the target is the simulator.

But let's stop this discussion, if you try I'm sure everything will be ok, these problems are just caused by my lack of knowledge in iOS development and Xcode.

**Comment by BlueCobold** [comment]
See, all my built SFML libraries for iOS are fat and iclude i386, x64, armv7, armv7s and arm64. However, like I said, I'm going to test it and report back. :)

**Comment by phoenixx2** [comment]
I'm not so sure we should merge this branch just yet.  sf::Sound doesn't seem to handle sf::SoundBuffer properly at all.  I've attached an audio file that demonstrates this issue, which I've tested with the "sound" example.  This is far from the only file affected.  In fact, most files don't play properly, including other file formats.  Sometimes they're cut short or sometimes they're completely corrupt.

All files work flawlessly through sf::Music, though.

http://puu.sh/ghVVw/aa40100ace.wav

**Comment by zsbzsb** [comment]
As my previous test was using the audio files in the examples they worked just fine. However I did just test the file @phoenixx2 provided and can confirm that this branch does not play it back in full. It cuts off halfway through playing. SFML reports the file as a total of .57 seconds in length - however when viewed in audacity the length is .94 seconds.

The only different I could see between this wav file and the ones in the examples is that this is stereo with a different bitrate.

**TL;DR** I can confirm with file provided by @phoenixx2 that this branch does not play the full file.

**Comment by LaurentGomila** [comment]
I've spotted the bug, I'll submit a fix ASAP.

Thanks for the feedback.

**Comment by LaurentGomila** [comment]
It's fixed, WAV files with more than 1 channel now play correctly.

> In fact, most files don't play properly, including other file formats.

This bug was really specific to the WAV reader. If you have files of other formats that don't play as expected, don't hesitate to upload them too.

**Comment by phoenixx2** [comment]
Rest are fine - my bad.  Misuse of `setRelativeToListener` on my own code.  Thanks for wav fix.

**Comment by zsbzsb** [comment]
Just wanted to point out before this is merged that there is still old copyright notices in the new header files that should be updated as in #805.

**Comment by BlueCobold** [comment]
Sounds play properly in iOS with the no_libsndfile_all_os branch.

But I get a EXC_BAD_ACCESS (aka segfault) on OSX in sf::~SoundBuffer though. There is one sound in the list and when resetBuffer gets called, the list suddenly turns empty which causes the iterator to turn invalid and causes a crash. Not cool. It happens every time when closing my game.

**Comment by LaurentGomila** [comment]
> Sounds play properly in iOS with the no_libsndfile_all_os branch.

:clap:

Simulator or real device?

**Comment by BlueCobold** [comment]
Real device.

**Comment by LaurentGomila** [comment]
> But I get a EXC_BAD_ACCESS (aka segfault) on OSX in sf::~SoundBuffer

This doesn't seem to be related to this branch. Can you test it with master?

**Comment by BlueCobold** [comment]
From checking the source in the master, the same issue exists there, so it is not related to this branch.

**Comment by LaurentGomila** [comment]
> From checking the source in the master, the same issue exists there, so it is not related to this branch.

Ok, so once the copyright notices are updated, can we consider that we are ready to merge?

**Comment by BlueCobold** [comment]
Looks fine on iOS and OS X then, if Windows, Linux and Android have been checked, I don't have any concerns.

**Comment by mantognini** [comment]
> So once the copyright notices are updated

Done.

**Comment by mantognini** [comment]
Conflict with master. Will fix it later.

**Comment by MarioLiebisch** [comment]
Fixed already - really just been two changes at the same position in the file.

**Comment by mantognini** [comment]
:+1:

**Comment by mantognini** [comment]
have all issues been fixed or am I mistaken?

**Comment by LaurentGomila** [comment]
Yes, I think we are good to go.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

**Comment by mantognini** [comment]
I've got some, sorry. :-P

I'm pretty much done reviewing it and just spot minor typos. Won't be long.

**Comment by mantognini** [comment]
Done. Nothing else for me.

**Comment by a-maly** [comment]
It is not possible to register your own file reader/writer that has been internally registered, just for the fact that you used to search forward instead of backward.
```
for (ReaderFactoryArray::const_iterator it = s_readers.begin(); it != s_readers.end(); ++it)
```

I'm just writing to let you know.

**Comment by MarioLiebisch** [comment]
@a-maly Hm... yeah, overwriting built-in filters sounds like a neat tidbit some might want to do.

**Comment by LaurentGomila** [comment]
I suggest that we release it now, get some feedback, and work more on it later if needed. Unless something is broken, this PR must be merged now :p

**Comment by mantognini** [comment]
I guess we could add a priority system, or something similar, in SFML 3 but I believe that for now this is enough. Most people will simply use the default implementation (which should be good enough in any case) I guess.

So let's proceed as-is. :-)

**Comment by eXpl0it3r** [comment]
:clap: :tada: :cocktail:

**Comment by mantognini** [comment]
> Now let's wait for all the bug reports.

Got one! :yum: 

It's a slight oversight that can lead to some hard-to-debug issues: `InputStream` is copyable. While it should be alright with the current implementation of `MemoryInputStream`, it is not good with `FileInputStream`.

There is one thing I'm not sure of: since `InputStream` prior to that merge was not exported, no one could be using it before, right?

If that's the case we can just make it inherit from `NonCopyable` and we're done. Otherwise we can make `FileInputStream` (and optionally `MemoryInputStream` too) inherit from `NonCopyable`.

**Comment by LaurentGomila** [comment]
```FileInputStream``` was not copyable <hen it used a ```std::ifsteam```, but now with ```FILE*``` it is indeed a problem. But there's no reason to disallow copy at the base class level, so let's just do it for ```FileInputStream```.

**Comment by mantognini** [comment]
Alright, will create a PR for that then.
